### PR TITLE
[Merge/Mux] Implement BasePad Time Synchronization Policy  - @open sesame 11/08 17:21

### PR DESCRIPTION
**Pull request by jijoongmoon**
# PR Description

Implement BasePad Time Synchronization Policy. The sync_mode name is
"basepad" and it has sync_options which is base pad number and
duration. It is more like "sync_mode=basepad
sync_option=0:33333333". Duration is GstClockTime (unsigned int 64)
and pad number is unsigned int.

Once this sync_mode is enabled, then the merge time is decided based
on this "basepad". For example, if the framerate of basepad is 30/1,
then the time which is pushed by merge/mux is 33333333, 66666666,
99999999, 133333333, 166666666, etc. Some details are below. It
calculate difference between current time and pts time of updated buffer 
and if it is bigger than duration, it use previous pad buffer.

Test Case 1:
 - srcpad0 ( framerate 30/1 ) : basepad with 33333333 duration.
 - srcpad1 ( framerate 20/1 )
 - srcpad2 ( framerate 10/1 )

then, merge/mux pushed as below.
````
    *srcpad0        srcpad1         srcpad2
       0               0               0     
    33333333        50000000           0     
    66666666        50000000           0    
    99999999       100000000      1000000000 
   133333332       150000000      1000000000
   166666666       150000000      1000000000
   199999999       200000000      2000000000
   233333331       250000000      2000000000
   266666666       250000000      2000000000 
   299999997       300000000      3000000000
````

if sync_mode is slowest,
then, merge/mux pushed as below.
````
    srcpad0         srcpad1         *srcpad2
       0               0               0     
    99999999       100000000      1000000000 
   199999999       200000000      2000000000
   299999997       300000000      3000000000 
````

if sync_mode is nosync,
then, merge/mux pushed as below.
````
    *srcpad0        *srcpad1        *srcpad2
       0               0               0     
    33333333        50000000      1000000000 
    66666666       100000000      2000000000 
    99999999       150000000      3000000000 
   133333332       200000000      4000000000 
   166666666       250000000      5000000000 
   199999999       300000000      6000000000
   233333331       350000000      7000000000 
   266666666       400000000      8000000000 
   299999997       450000000      9000000000 
````

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: gst/nnstreamer/tensor_common.c [hunk 1]
```diff
@@ -1132,17 +1133,21 @@ gst_tensor_set_current_time (GstCollectPads * collect,
...
         break;
       case SYNC_BASEPAD:
+        if (count == sync.data_basepad.sink_id)
+          *current_time = GST_BUFFER_PTS (buf);
+        gst_buffer_unref (buf);
         break;
       default:
         break;
     }
+    count++;
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_common.c]: unref-buffer
remove line 1140 and 1145, and add here 'gst_buffer_unref (buf);'

File: gst/nnstreamer/tensor_common.c [hunk 2]
```diff
@@ -1235,14 +1267,16 @@ gst_gen_tensors_from_collectpad (GstCollectPads * collect,
     if (GST_IS_BUFFER (buf)) {
       mem = gst_buffer_get_memory (buf, 0);
       gst_buffer_append_memory (tensors_buf, mem);
+      printf ("Append Memory # %d (PTS : %lu)\n",
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_common.c]: change func to print logs (printf -> silent_debug ?)

Comment (jijoongmoon) [reply gst/nnstreamer/tensor_common.c]: It is accidentally there. I will delete it. Thanks.

File: gst/nnstreamer/tensor_common.c (hunk 1 above)
Comment (jijoongmoon) [reply gst/nnstreamer/tensor_common.c]: right. I will change it.

**Comment by jijoongmoon** [comment]
I wrote synchronization policies in
https://github.com/nnsuite/nnstreamer/wiki/Synchronization-Policies-at-Mux-and-Merge

File: gst/nnstreamer/tensor_common.c [hunk 3]
```diff
@@ -1225,8 +1224,40 @@ gst_gen_tensors_from_collectpad (GstCollectPads * collect,
...
       case SYNC_NOSYNC:
-        gst_buffer_unref (buf);
-        buf = gst_collect_pads_pop (collect, data);
+        if (buf != NULL) {
+          gst_buffer_unref (buf);
+          buf = gst_collect_pads_pop (collect, data);
+        }
+        break;
+      case SYNC_BASEPAD:
+      {
```
Comment (ohsewon) [inline gst/nnstreamer/tensor_common.c]: This brace looks strange. Is it neccessary?

Comment (jijoongmoon) [reply gst/nnstreamer/tensor_common.c]: done.

**Pull request merged**